cuda_phase-retrieval_class/test/cupy_ref.hpp

```diff
@@ -0,0 +1,10 @@
+#pragma once
+
+template<typename TData>
+class Custom_Cupy_Ref {
```

Comment:
Add a second template parameter to fix the dimension:
```c++
template<typename TData, int TDim = 0>
class Custom_Cupy_Ref {
// ...
}
```
If `TDim` is 0, the dimension can be variable and will be not checked at casting the cupy_ref. If the dim is bigger than 0, check if the cupy_ref from the Python side has the correct dimension.